api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java

```diff
@@ -0,0 +1,92 @@
+import org.pmiops.workbench.model.RecentResourceResponse;
+import org.pmiops.workbench.firecloud.model.WorkspaceResponse;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.inject.Provider;
+import java.util.List;
+import java.util.function.Function;
+
+@RestController
+public class UserMetricsController implements UserMetricsApiDelegate {
+  Provider<User> userProvider;
+  UserRecentResourceService userRecentResourceService;
+  CohortService cohortService;
+  WorkspaceService workspaceService;
+  FireCloudService fireCloudService;
+
+  @Autowired
+  UserMetricsController(Provider<User> userProvider,
+      UserRecentResourceService userRecentResourceService,
+      CohortService cohortService,
+      WorkspaceService workspaceService,
+      FireCloudService fireCloudService) {
+    this.userProvider = userProvider;
+    this.userRecentResourceService = userRecentResourceService;
+    this.cohortService = cohortService;
+    this.workspaceService = workspaceService;
+    this.fireCloudService = fireCloudService;
+  }
+
+  /**
+   * Gets the list of all resources recently access by user in order of access date time
+   * @return
+   */
+  @Override
+  public ResponseEntity<RecentResourceResponse> getUserMetrics() {
+    long userId = userProvider.get().getUserId();
+    RecentResourceResponse response = new RecentResourceResponse();
+    List<UserRecentResource> userRecentResourceList = userRecentResourceService.findAllResourcesByUser(userId);
+
+    userRecentResourceList.stream().forEach(userRecentResource -> {
+      response.add(TO_CLIENT.apply(userRecentResource));
+    });
+
+    response.stream()
+        .filter(resource -> resource.getCohortId() != null)
+        .forEach(recentResource -> {
+          Cohort cohort = cohortService
+              .findCohortByWorkspaceIdAndCohortId(recentResource.getWorkspaceId(), recentResource.getCohortId());
+          recentResource.setName(cohort.getName());
+          recentResource.setDescription(cohort.getDescription());
+        });
+    return ResponseEntity.ok(response);
+  }
+
+  //Converts DB model to client Model
+  private final Function<UserRecentResource, RecentResource> TO_CLIENT =
+      userRecentResource -> {
+        RecentResource response = new RecentResource();
+        response.setCohortId(userRecentResource.getCohortId());
+        response.setModifiedTime(userRecentResource.getLastAccessDate().toString());
+        if (userRecentResource.getCohortId() == null) {
+          response.setName(userRecentResource.getNotebookName());
+          response.setType("notebook");
+          response.setDescription("");
+        } else {
+          response.setType("cohort");
+        }
+        response.setWorkspaceId(userRecentResource.getWorkspaceId());
+        Workspace workspace = workspaceService.findByWorkspaceId(response.getWorkspaceId());
```

Comment:
It's also potentially 10 DB calls to fetch cohort names, right? That's somewhat faster than 10 FC calls, but still slow.

Better would be to have either the cohort names, etc. denormalized into the resource entry (though this raises interesting questions about renames); or to do a fetch join in the query that retrieves the resource entries, so that the cohorts come back in the single SQL statement.

Reply:
We link to cohorts via cohort ID, but definitely agree that this would ideally be doing a batch cohort fetch instead.